src/hotspot/share/classfile/protectionDomainCache.cpp

```diff
@@ -128,108 +159,69 @@ void ProtectionDomainCacheTable::unlink() {
-  st->print_cr("Protection domain cache table (table_size=%d, classes=%d)",
-               table_size(), number_of_entries());
-  for (int index = 0; index < table_size(); index++) {
-    for (ProtectionDomainCacheEntry* probe = bucket(index);
-                                     probe != NULL;
-                                     probe = probe->next()) {
-      st->print_cr("%4d: protection_domain: " PTR_FORMAT, index, p2i(probe->object_no_keepalive()));
-    }
-  }
+  auto printer = [&] (WeakHandle& key, WeakHandle& value) {
+      st->print_cr("  protection_domain: " PTR_FORMAT, p2i(value.peek()));
+  };
+  st->print_cr("Protection domain cache table (table_size=%d, protection domains=%d)",
+                _pd_cache_table.table_size(), _pd_cache_table.number_of_entries());
+  _pd_cache_table.iterate_all(printer);
 }
 
 void ProtectionDomainCacheTable::verify() {
-  verify_table<ProtectionDomainCacheEntry>("Protection Domain Table");
-}
-
-oop ProtectionDomainCacheEntry::object() {
-  return literal().resolve();
+  auto verifier = [&] (WeakHandle& key, WeakHandle& value) {
+    guarantee(value.peek() == nullptr || oopDesc::is_oop(value.peek()), "must be an oop");
+  };
+  _pd_cache_table.iterate_all(verifier);
 }
 
 // The object_no_keepalive() call peeks at the phantomly reachable oop without
 // keeping it alive. This is okay to do in the VM thread state if it is not
 // leaked out to become strongly reachable.
-oop ProtectionDomainCacheEntry::object_no_keepalive() {
-  return literal().peek();
-}
-
 oop ProtectionDomainEntry::object_no_keepalive() {
-  return _pd_cache->object_no_keepalive();
+  return _object.peek();
 }
 
-void ProtectionDomainCacheEntry::verify() {
-  guarantee(object_no_keepalive() == NULL || oopDesc::is_oop(object_no_keepalive()), "must be an oop");
-}
-
-ProtectionDomainCacheEntry* ProtectionDomainCacheTable::get(Handle protection_domain) {
-  unsigned int hash = compute_hash(protection_domain);
-  int index = hash_to_index(hash);
-
-  ProtectionDomainCacheEntry* entry = find_entry(index, protection_domain);
-  if (entry == NULL) {
-    entry = add_entry(index, hash, protection_domain);
-  }
-  // keep entry alive
-  (void)entry->object();
-  return entry;
-}
-
-ProtectionDomainCacheEntry* ProtectionDomainCacheTable::find_entry(int index, Handle protection_domain) {
+WeakHandle ProtectionDomainCacheTable::add_if_absent(Handle protection_domain) {
   assert_locked_or_safepoint(SystemDictionary_lock);
-  for (ProtectionDomainCacheEntry* e = bucket(index); e != NULL; e = e->next()) {
-    if (e->object_no_keepalive() == protection_domain()) {
-      return e;
+  WeakHandle w(Universe::vm_weak(), protection_domain);
+  bool created;
+  WeakHandle* wk = _pd_cache_table.put_if_absent(w, w, &created);
+  if (!created) {
+    // delete the one created since we already had it in the table
+    w.release(Universe::vm_weak());
+    // Keep entry alive
+    (void)wk->resolve();
+    return *wk;
```

Comment:
Can't this be factored out of the if-else as `wk` is always the right entry to resolve and return?

Reply:
Yes, that works.